misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt

```diff
@@ -0,0 +1,381 @@
+      if (newScatterQueryCount > count.get()) {
+        throw FullScatterException(
+            "Query scattered to all shards. This is expensive and prevents scalability because " +
+                "we won't be able to decrease load on each shard by shard splitting. Please " +
+                "introduce a lookup table vindex. Query was: $query")
+      }
+    }
+  }
+
+  inner class CowriteDetector : ExtendedQueryExectionListener() {
+    val keyspaceIdsThisTransaction: ThreadLocal<HashSet<String>> =
+        ThreadLocal.withInitial { HashSet<String>() }
+
+    override fun beforeStartTransaction() {
+      // Connect before the query because connecting spits out a bunch of crap in the general_log
+      // that makes it harder for us to get to the thread id
+      connect()
+
+      check(keyspaceIdsThisTransaction.get().isEmpty()) {
+        "Transaction state has not been cleaned up, beforeEndTransaction was never executed"
+      }
+    }
+
+    override fun afterQuery(query: String) {
+      if (!isDml(query)) return
+
+      val queryInDatabase = extractLastDmlQuery() ?: return
+
+      val m = "vtgate:: keyspace_id:([^ ]+)".toRegex().find(queryInDatabase) ?: return
+      val keyspaceId = m.groupValues[1]
+
+      val keyspaceIds = keyspaceIdsThisTransaction.get()
+      keyspaceIds.add(keyspaceId)
+
+      if (keyspaceIds.size > 1) {
+        throw CowriteException(
+            "DML against more than one entity group in the same transaction. " +
+                "These are not guaranteed to be ACID across shard splits and should be avoided. " +
+                "Query was: $queryInDatabase")
+      }
+    }
+
+    override fun beforeEndTransaction() {
+      keyspaceIdsThisTransaction.get().clear()
+    }
+  }
+
+  val COMMENT_PATTERN = "/\\*+[^*]*\\*+(?:[^/*][^*]*\\*+)*/".toRegex()
+  val DML = hashSetOf("insert", "delete", "update")
+
+  private fun isDml(query: String): Boolean {
+    val first = query
+        .replace(COMMENT_PATTERN, "")
+        .trimStart()
+        .toLowerCase()
+        .takeWhile { !it.isWhitespace() }
+    return DML.contains(first)
+  }
+
+  /**
+   * Digs into the MySQL log to find the last executed DML statement that passed through Vitess.
+   */
+  private fun extractLastDmlQuery(): String? {
+    return connect().let { c ->
+      c.createStatement().use { s ->
+        // OFFSET 1 because this query itself is also directly logged before it's executed
+        // "/* _stream" is appended to every DML statement by Vitess so that we skip a bunch of
+        // irrelevant statements
+        s.executeQuery("""
+                  SELECT argument
+                  FROM mysql.general_log
+                  WHERE command_type = 'Query'
+                  AND argument LIKE '%/* _stream%'
+                  ORDER BY event_time DESC
+                  LIMIT 1 OFFSET 1
+                """.trimIndent())
+            .uniqueResult { it.getString(1) }
```

Comment:
This is admittedly pretty clunky. I think the right way to do this is to have a metric in Vitess which exposes how many keyspace IDs we encountered. We check the value before each DML statement and then again after each DML statement, you can only make it go up by one tick per transaction.

That way we:
1) don't have to care about how Vitess rewrites queries.
2) can use the same metric to monitor for cowrites in transction, if we divide total entity groups written to by total transactions we will get entity groups per transaction which should never go above 1.

I'm working on such a metric now.

Reply:
Actually I'm not working on it. I'm gonna switch to working on the deployment side of things next week instead. This will have to do for now.